File: apps/api/internal/api/hservices/dto.go
package hservices

import "github.com/jackc/pgx/v5/pgtype"

type CreateHServiceRequestDto struct {
	Title            string  `json:"title" validate:"required,min=5,max=64"`
	Description      string  `json:"description" validate:"required,min=5,max=4096"`
	Category         int     `json:"category" validate:"required,min=1,max=32"`
	Price            float64 `json:"price" validate:"required,min=1,max=10000"`
	PriceUnit        string  `json:"priceUnit" validate:"required"`
	PriceTimespan    string  `json:"priceTimespan" validate:"required"`
	IsOnline         *bool   `json:"isOnline" validate:"required"`
	Url              *string `json:"url"`
	Location         string  `json:"location" validate:"required,min=1,max=256"`
	DeliveryTime     int     `json:"deliveryTime" validate:"required,min=1,max=32"`
	DeliveryTimespan string  `json:"deliveryTimespan" validate:"required"`
	Media            string  `json:"media"`
}

type HServiceResponseDto struct {
	ID               string             `json:"id"`
	UserID           string             `json:"userId"`
	Title            string             `json:"title"`
	Slug             string             `json:"slug"`
	Description      string             `json:"description"`
	Category         int32              `json:"category"`
	Price            float64            `json:"price"`
	PriceUnit        string             `json:"priceUnit"`
	PriceTimespan    string             `json:"priceTimespan"`
	IsOnline         bool               `json:"isOnline"`
	Url              pgtype.Text        `json:"url"`
	Location         string             `json:"location"`
	DeliveryTime     int32              `json:"deliveryTime"`
	DeliveryTimespan string             `json:"deliveryTimespan"`
	TotalPoints      int64              `json:"totalPoints"`
	TotalVotes       int32              `json:"totalVotes"`
	Media            map[string]any     `json:"media"`
	CreatedAt        pgtype.Timestamptz `json:"createdAt"`
	UpdatedAt        pgtype.Timestamptz `json:"updatedAt"`
}

File: apps/web/src/lib/dto.ts
export type GetMeResponseDto = {
  id: string;
  email: string;
  username: string;
  fullName: string;
  googleId: string | null;
  isEmailVerified: boolean;
  isActive: boolean;
  role: string;
  gender: string | null;
  profileImage: string | null;
  lastLogin: string;
  createdAt: string;
  updatedAt: string;
};

export type UploadImageType = 'hservices' | 'profile' | 'reviews';

export type CreateHServiceRequestDto = {
  title: string;
  description: string;
  category: number;
  price: number;
  priceUnit: string;
  priceTimespan: string;
  isOnline: boolean;
  url: string;
  location: string;
  deliveryTime: number;
  deliveryTimespan: string;
  media: string;
};

export type UserDto = GetUserProfileByUsernameResponseDto;

export type HServiceResponseDto = {
  id: string;
  user: UserDto;
  userId: string;
  title: string;
  slug: string;
  description: string;
  category: number;
  price: number;
  priceUnit: string;
  priceTimespan: string;
  isOnline: boolean;
  url: string | null;
  location: string;
  deliveryTime: number;
  deliveryTimespan: string;
  totalPoints: number;
  totalVotes: number;
  media: MediaResponseDto;
  createdAt: string;
  updatedAt: string;
};

export type HServiceWithoutUserResponseDto = {
  id: string;
  userId: string;
  title: string;
  slug: string;
  description: string;
  category: number;
  price: number;
  priceUnit: string;
  priceTimespan: string;
  isOnline: boolean;
  url: string | null;
  location: string;
  deliveryTime: number;
  deliveryTimespan: string;
  totalPoints: number;
  totalVotes: number;
  media: MediaResponseDto;
  createdAt: string;
  updatedAt: string;
};

export type MediaResponseDto = {
  data: Media[];
};

export type Media = {
  type: 'image' | 'video';
  url: string;
  thumbnail: string;
  alt: string;
  caption: string | null;
  width: number;
  height: number;
};

export type GetUserProfileByUsernameResponseDto = {
  id: string;
  username: string;
  fullName: string;
  gender: string | null;
  profileImage: string | null;
  createdAt: string | null;
};

export type GetNewUploadUrlResponseDto = {
  Key: string;
  Url: string;
};
